shared/actions/signup/index.js

```diff
@@ -156,219 +83,344 @@ function isValidName (name: ?string): ?string {
+    payload: {paperkey: new HiddenString(paperKey.payload.phrase)}
+  })
+  yield put(routeAppend('paperkey'))
+  yield take(Constants.sawPaperKey)
+  paperKey.payload.response.result()
 }
 
-let paperKeyResponse = null
-export function sawPaperKey (): AsyncAction {
-  return () => {
-    if (paperKeyResponse) {
-      paperKeyResponse.result()
-      paperKeyResponse = null
-    }
+function * signupFinish (): any {
+  // $FlowIssue
+  const {payload: {error}} = yield take('keybase.1.signup.signup')
+  if (error) {
+    yield put({type: Constants.signup, error: true, payload: {signupError: new HiddenString(error + '')}})
+    yield put(routeAppend('signupError'))
+    return
   }
+  console.log('Successful signup')
 }
 
-function signup (skipMail: boolean, onDisplayPaperKey?: () => void): TypedAsyncAction<Signup | ShowPaperKey | RouteAppend | SignupWaiting> {
-  return (dispatch, getState) => new Promise((resolve, reject) => {
-    const {email, username, inviteCode, passphrase, deviceName} = getState().signup
-    paperKeyResponse = null
-
-    if (email && username && inviteCode && passphrase && deviceName) {
-      const params: signup_signup_rpc = {
-        method: 'signup.signup',
-        waitingHandler: isWaiting => dispatch(waiting(isWaiting)),
-        param: {
-          email,
-          inviteCode,
-          username,
-          deviceName,
-          passphrase: passphrase.stringValue(),
-          storeSecret: false,
-          skipMail
-        },
-        incomingCallMap: {
-          'keybase.1.loginUi.displayPrimaryPaperKey': ({sessionID, phrase}, response) => {
-            paperKeyResponse = response
-            dispatch({
-              type: Constants.showPaperKey,
-              payload: {paperkey: new HiddenString(phrase)}
-            })
-            onDisplayPaperKey && onDisplayPaperKey()
-            dispatch(nextPhase())
-          },
-          'keybase.1.gpgUi.wantToAddGPGKey': (params, {error, result}) => {
-            // Do not add a gpg key for now
-            result(false)
-          }
-        },
-        callback: (err, {passphraseOk, postOk, writeOk}) => {
-          if (err) {
-            console.error('error in signup:', err)
-            dispatch({
-              type: Constants.signup,
-              error: true,
-              payload: {signupError: new HiddenString(err + '')}
-            })
-            dispatch(nextPhase())
-            reject()
-          } else {
-            console.log('Successful signup', passphraseOk, postOk, writeOk)
-            resolve()
-          }
-        }
-      }
+function signupArgsSelector (state) {
+  const {email, username, inviteCode, passphrase, deviceName} = state.signup
+  return {email, username, inviteCode, passphrase, deviceName}
+}
 
-      engine.rpc(params)
-    } else {
-      console.error('Entered signup action with a null required field')
-      reject()
+// The real meat, this describes how the signup flow
+function * signupFlow (): any {
```

Comment:
This is pretty cool, We are describing the flow and each screen defer to the saga that deals with that screen.